solanaj/src/main/java/org/p2p/solanaj/ws/SubscriptionWebSocketClient.kt

```diff
@@ -0,0 +1,143 @@
+        val rpcRequest = RpcRequest("accountSubscribe", mutableListOf(key))
+        subscriptions[rpcRequest.id] = SubscriptionParams(rpcRequest, listener)
+        subscriptionIds[rpcRequest.id] = null
+        updateSubscriptions()
+    }
+
+    fun signatureSubscribe(signature: String, listener: NotificationEventListener) {
+        val rpcRequest = RpcRequest("signatureSubscribe", mutableListOf(signature))
+        subscriptions[rpcRequest.id] = SubscriptionParams(rpcRequest, listener)
+        subscriptionIds[rpcRequest.id] = null
+        updateSubscriptions()
+    }
+
+    override fun onWebsocketPong(socket: WebSocket, framedata: Framedata) {
+        super.onWebsocketPong(socket, framedata)
+        socketStateListener?.onWebSocketPong()
+    }
+
+    override fun onOpen(handshakedata: ServerHandshake) {
+        socketStateListener?.onConnected()
+        updateSubscriptions()
+    }
+
+    override fun onMessage(message: String) {
+        Log.d("SOCKET", "New message received: $message")
+        val resultAdapter = Moshi.Builder().build()
+            .adapter<RpcResponse<Long>>(Types.newParameterizedType(RpcResponse::class.java, Long::class.java))
+        try {
+            resultAdapter.fromJson(message)?.also { rpcResult ->
+                val rpcResultId = rpcResult.id
+                if (rpcResultId != null) {
+                    if (subscriptionIds.containsKey(rpcResultId)) {
+                        subscriptionIds[rpcResultId] = rpcResult.result
+                        subscriptions[rpcResultId]?.listener?.let { listener ->
+                            subscriptionListeners[rpcResult.result] = listener
+                        }
+                        subscriptions.remove(rpcResultId)
+                    }
+                } else {
+                    val notificationResultAdapter = Moshi.Builder().build().adapter(RpcNotificationResult::class.java)
+                    notificationResultAdapter.fromJson(message)?.let { result ->
+                        val listener = subscriptionListeners[result.params.subscription]
+                        val value = result.params.result.value as Map<*, *>
+                        when (result.method) {
+                            "signatureNotification" -> listener!!.onNotificationEvent(
+                                SignatureNotification(
+                                    value["err"]
+                                )
+                            )
+                            "accountNotification" -> listener!!.onNotificationEvent(value)
+                        }
+                    }
+                }
+            }
+        } catch (ex: Exception) {
+            Log.e("SOCKET", "Error on socket message", ex)
+        }
+    }
+
+    override fun onClose(code: Int, reason: String, remote: Boolean) {
+        socketStateListener?.onClosed(
+            code,
+            "Connection closed by "
```

Comment:
that's a lot of .plus operators here. Joining strings this way is bad - remember that Strings are immutable in Java. It's better to rewrite this part using buildString

Reply:
controversial moment, is the game worth the candle